File: programmers/힙_더맵게.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(vector<int> scoville, int K) {
    int answer = 0;
    
    priority_queue<int, vector<int>, greater<int> > pq;
    
    for (auto s : scoville) {
        pq.push(s);
    }
    
    while (pq.size() > 1 && pq.top() < K) {
        int first = pq.top();
        pq.pop();

        int second = pq.top();
        pq.pop();

        if (pq.size() == 0) {
            // 원소가 두개인 경우
            if (first + second * 2 < K) {
                return -1;
            }
        }

        pq.push(first + second * 2);
        answer += 1;      
    }
    
    return answer;
}

File: programmers/힙_더맵게.py
import heapq

# priority queue 만들기 (default: min_heap)
heap = []

def solution(scoville, K):
    answer = 0
    
    if len(scoville) < 2:
        if scoville[0] < K:
            return -1
        else:
            return 0
    
    # 리스트를 우선순위큐(min_heap)으로 변경
    # O(n)
    heapq.heapify(scoville)
    
    while len(scoville) >= 1:
        top = heapq.heappop(scoville)

        if top >= K:
            break;
        elif len(scoville) == 0 and top < K:
            answer = -1
            break
            
        top2 = heapq.heappop(scoville)
        heapq.heappush(scoville, top + top2 * 2)
        answer += 1

    
    
    return answer